### PR TITLE
[sysrst_ctrl,sival] Various improvements

### PR DESCRIPTION
**Pull request by pamaury**
Read the commits for a description. In short:
- fix `ulp_z3_wakeup`
- introduce gpio monitor test util and use it in tests

### CODE REVIEW

**Review by jwnrt — APPROVED** [review]
This LGTM, thanks especially for making that nice VCD dumping module. Regarding sending the test phase over GPIO, that looks like it will work quite nicely but FYI I've seen other tests use the reset reason from the `rstmgr` to restore the correct test phase when reset during the test. I'm not sure whether that works for deep sleep, but your solution is good as well.

File: sw/host/tests/chip/sysrst_ctrl/sysrst_ctrl_ulp_z3_wakeup.rs [hunk 1]
```diff
@@ -86,45 +110,95 @@ fn chip_sw_sysrst_ctrl_ulp_z3_wakeup(
...
     log::info!("Go to PhaseDone");
-    MemWriteReq::execute(uart, test_phase_addr, &[TestPhase::Done as u8])?;
+    set_test_phase(transport, TestPhase::Done)?;
+
+    // Stop monitoring and check that there were no glitches.
+    let events = gpio_mon.read(false)?;
+    let mut events_iter = events.iter();
+    let Some(first_event) = events_iter.next() else {
+        bail!("Expected at least one GPIO event during reset");
+    };
```
Comment (engdoreis) [inline sw/host/tests/chip/sysrst_ctrl/sysrst_ctrl_ulp_z3_wakeup.rs:170]:
```suggestion
    let first_event = events_iter.next().context("Expected at least one GPIO event during reset");
```

**Review by engdoreis — APPROVED** [review]
LGTM

File: sw/device/tests/sysrst_ctrl_ulp_z3_wakeup_test.c [hunk 2]
```diff
@@ -102,6 +106,18 @@ static void pinmux_setup(void) {
...
+  sysrst_ctrl_testutils_release_dio(&sysrst_ctrl, true, true);
+}
+
+static uint8_t read_phase_pins(void) {
+  uint32_t gpio_state;
+  CHECK_DIF_OK(dif_gpio_read_all(&gpio, &gpio_state));
+  // Since the host may not be able to change all pins atomically, we use
+  // a Gray code encoding so that it suffices to change one pin to go to
+  // the next phase.
+  static const uint8_t kGrayCode[] = {0, 1, 3, 2, 7, 6, 4, 5};
```
Comment (jettr) [inline sw/device/tests/sysrst_ctrl_ulp_z3_wakeup_test.c:119]: Note the jump between `2` and `7` is 2 bits, the graycode should be:

`static const uint8_t kGrayCode[] = {0, 1, 3, 2, 6, 7, 5, 4};`

to only flip one bit at a time

Comment (pamaury) [reply sw/device/tests/sysrst_ctrl_ulp_z3_wakeup_test.c:119]: This is for decoding the Gray code, not encoding, so it's the "reversed" mapping.

Comment (jettr) [reply sw/device/tests/sysrst_ctrl_ulp_z3_wakeup_test.c:119]: Ah, I see. You might want to add a comment saying that it is an inverse look up, but the code it correct. Thanks!

**Pull request merged**